File: 2025/python/solve_day4_part1.py
def solve_day4_part1(text):
    nrows = len(text)
    ncols = len(text[0])
    liftable_count = 0

    for row in range(nrows):
        for col in range(ncols):
            if is_liftable(text, row, col, nrows, ncols):
                liftable_count += 1

    return(liftable_count)

def is_liftable(grid, row, col, nrows, ncols):
    char = grid[row][col]
    roll_count = 0

    if char != "@":
        return(False)

    for row_adj in range(-1, 2):
        search_row = row + row_adj
        if search_row < 0 or search_row >= nrows:
            continue

        for col_adj in range(-1, 2):
            search_col = col + col_adj

            if search_col < 0 or search_col >= ncols:
                continue
            if row == search_row and col == search_col:
                continue

            if grid[search_row][search_col] == "@":
                roll_count += 1

    return(roll_count < 4)

File: 2025/R/solve_day4_part1.R
solve_day4_part1 <- function(input) {
  grid <- aoc_input_to_chr_matrix(input)
  nrows <- NROW(grid)
  ncols <- NCOL(grid)
  liftable <- matrix(FALSE, nrows, ncols)

  for (row in seq_len(nrows)) {
    for (col in seq_len(ncols)) {
      liftable[row, col] <- is_liftable(grid, row, col, nrows, ncols)
    }
  }

  sum(liftable)
}

is_liftable <- function(grid, row, col, nrows, ncols) {
  char <- grid[row, col]
  if (char != "@") {
    return(FALSE)
  }

  # count surrounding rolls
  roll_count <- 0L

  for (row_adj in -1:1) {
    search_row <- row + row_adj
    if ((search_row) < 1) next
    if ((search_row) > nrows) next

    for (col_adj in -1:1) {
      search_col <- col + col_adj
      if ((search_col) < 1) next
      if ((search_col) > ncols) next
      if (row == search_row && col == search_col) next

      if (grid[search_row, search_col] == "@") {
        roll_count <- roll_count + 1L
      }
    }
  }

  roll_count < 4
}
